File: 05-Model Selection/ex5.py
from __future__ import division
import numpy as np
from scipy import optimize, io
import matplotlib.pyplot as plt
from sklearn.preprocessing import PolynomialFeatures

def cost_function(theta, X, y, lamda):
    m, n = X.shape
    J = np.sum(np.power((X.dot(theta.T)-y),2))/(2*m)  + (lamda/(2*m))*np.sum(np.square(theta[1:]))
    mask = np.eye(len(theta))
    mask[0, 0] = 0
    grad = (1/m)*(X.dot(theta.T) - y).T.dot(X) + (lamda/m)*mask.dot(theta)
    return J, grad

def theta_estimation(X, y, lamda):
    m, n = X.shape
    initial_theta = np.zeros((n, 1))
    result = optimize.minimize(fun=cost_function, x0=initial_theta, args=(X, y, lamda),
                               method='TNC', jac=True, options={'maxiter': 100})
    theta = result.x
    return theta

def learning_curve(X, y, X_val, y_val, lamda):
    m, n = X.shape
    m_val, n_val = X_val.shape
    error_train = np.zeros(m)
    error_val = np.zeros(m)

    for i in range(1, m + 1):
        theta = theta_estimation(X[:i,], y[:i,], lamda)
        error_train[i-1] = (1/(2*i))*np.sum(np.square(X[:i,].dot(theta) - y[:i,]))
        error_val[i-1] = (1/(2*m_val))*np.sum(np.square(X_val.dot(theta) - y_val))
    return error_train, error_val

def lamda_estimation(X, y, X_val, y_val):
    lambda_vec = np.array([0, 0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3, 10])
    m, n = X.shape
    m_val, n_val = X_val.shape
    error_train = np.zeros(len(lambda_vec))
    error_val = np.zeros(len(lambda_vec))

    for i in range(len(lambda_vec)):
        theta = theta_estimation(X, y, lambda_vec[i])
        error_train[i] = (1/(2*m))*np.sum(np.square(X.dot(theta) - y))
        error_val[i] = (1/(2*m_val))*np.sum(np.square(X_val.dot(theta) - y_val))
    return lambda_vec, error_train, error_val

def degree_estimation(data, lamda):
    degree_vec = np.array([2, 3, 4, 5, 6, 7, 8])
    y = data['y'].ravel()
    y_val = data['yval'].ravel()
    error_train = np.zeros(len(degree_vec))
    error_val = np.zeros(len(degree_vec))

    for i in range(len(degree_vec)):
        X_poly, X_val_poly, X_test_poly = poly_features(data, degree_vec[i])
        m, n = np.shape(X_poly)
        m_val, n_val = X_val_poly.shape
        theta = theta_estimation(X_poly, y, lamda)
        error_train[i] = (1/(2*m))*np.sum(np.square(X_poly.dot(theta) - y))
        error_val[i] = (1/(2*m_val))*np.sum(np.square(X_val_poly.dot(theta) - y_val))
    return degree_vec, error_train, error_val

def feature_normalize(X):
    mu = np.mean(X, axis=0)
    sigma = np.std(X, ddof=1, axis=0)
    X_norm = (X - mu) / sigma
    return X_norm

def poly_features(data,d):
    X = data['X']
    X_test = data['Xtest']
    X_val = data['Xval']

    poly = PolynomialFeatures(degree=d)
    X_poly = poly.fit_transform(X)
    X_poly = X_poly[:,1:]
    X_poly = feature_normalize(X_poly)
    m, n = X.shape
    X_poly = np.hstack((np.ones((m, 1)), X_poly))

    X_val = poly.fit_transform(X_val)
    X_val = X_val[:, 1:]
    X_val = feature_normalize(X_val)
    m_val, n_val = X_val.shape
    X_val = np.hstack((np.ones((m_val, 1)), X_val))

    X_test = poly.fit_transform(X_test)
    X_test = X_test[:, 1:]
    X_test = feature_normalize(X_test)
    m_test, n_test = X_test.shape
    X_test = np.hstack((np.ones((m_test, 1)), X_test))
    return X_poly, X_val, X_test

def plot_error_for_features(error_train, error_val):
    m = len(error_train)
    plt.plot(range(1, m + 1), error_train, color='b', label='Train')
    plt.plot(range(1, m + 1), error_val, color='r', label='Cross Validation')
    plt.legend(loc='upper right')
    plt.xlabel('Number of training examples')
    plt.ylabel('Error')

def plot_error(parameter, error_train, error_val):
    plt.plot(parameter, error_train, color='b', label='Train')
    plt.plot(parameter, error_val, color='r', label='Cross Validation')
    plt.legend(loc='upper right')
    plt.ylabel('Error')

if __name__ == "__main__":

    data = io.loadmat('ex5data1.mat')
    X = data['X']
    y = data['y'].ravel()
    X_test = data['Xtest']
    y_test = data['ytest'].ravel()
    X_val = data['Xval']
    y_val = data['yval'].ravel()

    m, n = X.shape
    X = np.hstack((np.ones((m, 1)), X))
    inital_theta = np.ones(n + 1)
    lamda = 1

    J, grad = cost_function(inital_theta, X, y, lamda)
    theta = theta_estimation(X, y, lamda)
    pred = X.dot(theta.T)

    m_val, n_val = X_val.shape
    X_val = np.hstack((np.ones((m_val, 1)), X_val))
    error_train, error_val = learning_curve(X, y, X_val, y_val, lamda)

    plt.figure(1)
    plot_error_for_features(error_train, error_val)

    degree = 5
    X_poly, X_val_poly, X_test_poly = poly_features(data, degree)
    m, n = np.shape(X_poly)
    initial_theta = np.zeros((n, 1))
    theta_poly = theta_estimation(X_poly, y, lamda)
    pred_poly = X_poly.dot(theta_poly.T)

    error_train_poly, error_val_poly = learning_curve(X_poly, y, X_val_poly, y_val, lamda)
    plt.figure(2)
    plot_error_for_features(error_train_poly, error_val_poly)

    lambda_vec, error_train_lamda, error_val_lamda = lamda_estimation(X_poly, y, X_val_poly, y_val)
    plt.figure(3)
    plot_error(lambda_vec, error_train_lamda, error_val_lamda)
    plt.xlabel('lambda')

    degree_vec, error_train_degree, error_val_degree = degree_estimation(data, lamda)
    plt.figure(4)
    plot_error(degree_vec, error_train_degree, error_val_degree)
    plt.xlabel('Degree')

    plt.figure(5)
    plt.plot(X[:,1], y, linestyle='', marker='x', color='r', label='Original')
    plt.plot(X[:,1], pred, linestyle='--', marker='o', color='b', label='Linear Fit')
    plt.plot(X[:,1], pred_poly, linestyle='', marker='*', color='g', label='Poly Fit')
    plt.legend(loc='upper right')
    plt.xlabel('Change in water level (x)')
    plt.ylabel('Water flowing out of the dam (y)')
    plt.show()

File: 05-Model Selection/ex5.m
load ('ex5data1.mat');
m = size(X, 1);
plot(X, y, 'rx', 'MarkerSize', 10, 'LineWidth', 1.5);
xlabel('Change in water level (x)');
ylabel('Water flowing out of the dam (y)');

function [J, grad] = linearRegCostFunction(X,y, theta, lambda)
  m=length(y);
  J = 0;
  grad = zeros(size(theta));
  theta_zeroed_first = [0; theta(2:length(theta));];
  J = (1/(2*m))*sumsq((X * theta) - y) + (lambda/(2*m))*sumsq(theta_zeroed_first);
  for (j = 1:length(grad))
    grad(j) = (1/m)*sum(((X * theta) - y).* X(:,j)) + ((lambda/m)*theta_zeroed_first(j));
  endfor
endfunction

function [theta] = trainLinearReg(X, y, lambda)
  initial_theta = zeros(size(X, 2), 1); 
  costFunction = @(t) linearRegCostFunction(X, y, t, lambda);
  options = optimset('MaxIter', 200, 'GradObj', 'on');
  theta = fmincg(costFunction, initial_theta, options);
endfunction

function [error_train, error_val] = learningCurve(X, y, Xval, yval, lambda)
  m = size(X, 1);
  mval = size(Xval, 1);
  error_train = zeros(m, 1);
  error_val   = zeros(m, 1);
  for i = 1:m
  	theta = trainLinearReg(X(1:i, :), y(1:i), lambda);
	  error_train(i) = (1/(2*i))*sumsq((X(1:i,:)*theta) - y(1:i));
	  error_val(i) = (1/(2*mval))*sumsq((Xval * theta) - yval);
  endfor
endfunction

function [X_poly] = polyFeatures(X, p)
  X_poly = zeros(numel(X), p);
  for i = 1:p
	  X_poly(:, i) = X .^ i;
  endfor
endfunction

function [X_norm, mu, sigma] = featureNormalize(X)
  X_norm = zeros(size(X));
  mu = zeros(1, size(X, 2));
  sigma = zeros(1, size(X, 2));
  for i = 1:size(X,2)
    mu(i) = mean(X(:,i));
    sigma(i) = std(X(:,i));
    for j = 1:size(X,1) 
	    X_norm(j,i) = (X(:,i)(j) - mu(i))/sigma(i);
    endfor
  endfor
endfunction

function [lambda_vec, error_train_vec, error_val_vec] = validationCurve(X_poly, y, X_poly_val, yval)
  m = size(X_poly, 1);
  mpoly = size(X_poly_val, 1);
  lambda_vec = [0, 0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1, 3, 10]
  error_train_vec = zeros(size(lambda_vec), (2));
  error_val_vec   = zeros(size(lambda_vec), (2));  
  for i = 1:length(lambda_vec)
  	theta = trainLinearReg(X_poly, y, lambda_vec(i));
	  error_train_vec(i) = (1/(2*m))*sumsq((X_poly * theta) - y);
	  error_val_vec(i) = (1/(2*mpoly))*sumsq((X_poly_val * theta) - yval);
  endfor    
endfunction

X = [ones(m, 1) X];
theta = [1 ; 1];
lambda = 1;
[J, grad] = linearRegCostFunction(X, y, theta, lambda);
[theta] = trainLinearReg(X, y, lambda);

hold on;
plot(X(:,2), X*theta, '-', 'LineWidth', 2)
hold off;

mval = size(Xval, 1);
Xval = [ones(mval, 1) Xval];
lambda = 0;
[error_train, error_val] = learningCurve(X, y, Xval, yval, lambda);

plot(1:m, error_train, 1:m, error_val);
title('Learning curve for linear regression')
legend('Train', 'Cross Validation')
xlabel('Number of training examples')
ylabel('Error')
axis([0 13 0 150])

p = 8;
X_poly = polyFeatures(X(:,2), p);
[X_poly, mu, sigma] = featureNormalize(X_poly);
X_poly = [ones(m, 1) X_poly];
lambda = 1;
[theta_poly] = trainLinearReg(X_poly, y, lambda);

X_poly_test = polyFeatures(Xtest, p);
X_poly_test = bsxfun(@minus, X_poly_test, mu);
X_poly_test = bsxfun(@rdivide, X_poly_test, sigma);
X_poly_test = [ones(size(X_poly_test, 1), 1), X_poly_test];       

X_poly_val = polyFeatures(Xval(:,2), p);
X_poly_val = bsxfun(@minus, X_poly_val, mu);
X_poly_val = bsxfun(@rdivide, X_poly_val, sigma);
X_poly_val = [ones(size(X_poly_val, 1), 1), X_poly_val];           

figure(1);
plot(X(:,2), y, 'rx', 'MarkerSize', 10, 'LineWidth', 1.5);
plotFit(min(X(:,2)), max(X(:,2)), mu, sigma, theta_poly, p);
xlabel('Change in water level (x)');
ylabel('Water flowing out of the dam (y)');
title (sprintf('Polynomial Regression Fit (lambda = %f)', lambda));

figure(2);
[error_train, error_val] = learningCurve(X_poly, y, X_poly_val, yval, lambda);
plot(1:m, error_train, 1:m, error_val);

title(sprintf('Polynomial Regression Learning Curve (lambda = %f)', lambda));
xlabel('Number of training examples')
ylabel('Error')
axis([0 13 0 100])
legend('Train', 'Cross Validation')

fprintf('Polynomial Regression (lambda = %f)\n\n', lambda);
fprintf('# Training Examples\tTrain Error\tCross Validation Error\n');
for i = 1:m
    fprintf('  \t%d\t\t%f\t%f\n', i, error_train(i), error_val(i));
end

[lambda_vec, error_train_vec, error_val_vec] = validationCurve(X_poly, y, X_poly_val, yval);
plot(lambda_vec, error_train_vec, lambda_vec, error_val_vec);
legend('Train', 'Cross Validation');
xlabel('lambda');
ylabel('Error');

fprintf('lambda\t\tTrain Error\tValidation Error\n');
for i = 1:length(lambda_vec)
	fprintf(' %f\t%f\t%f\n', lambda_vec(i), error_train_vec(i), error_val_vec(i));
end
